pca_cca.go

```diff
@@ -0,0 +1,262 @@
+		f = 1 / (floats.Sum(weights) - 1)
+	}
+	for i, v := range vars {
+		vars[i] = f * v * v
+	}
+	return vecs, vars, true
+}
+
+// CC is the result of a canonical correlation analysis. A CC should not be used
+// unless it has been returned by a successful call to CanonicalCorrelations.
+type CC struct {
+	// n is the number of observations used to
+	// construct the canonical correlations.
+	n int
+
+	x, y, c *mat64.SVD
+}
+
+// Corrs returns the canonical correlations in dest.
+func (c CC) Corrs(dest []float64) []float64 {
+	if c.c == nil {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	return c.c.Values(dest)
+}
+
+// Left returns the left-eigenvectors of the canonical correlation matrix if
+// spheredSpace is true, if spheredSpace is false it returns these eigenvectors
+// back-transformed to the original data space.
+func (c CC) Left(spheredSpace bool) (vecs *mat64.Dense) {
+	if c.c == nil || c.x == nil || c.n < 2 {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	var pv mat64.Dense
+	pv.UFromSVD(c.c)
+	if spheredSpace {
+		return &pv
+	}
+
+	var xv mat64.Dense
+	xs := c.x.Values(nil)
+	xv.VFromSVD(c.x)
+
+	scaleColsReciSqrt(&xv, xs)
+
+	pv.Product(&xv, xv.T(), &pv)
+	pv.Scale(math.Sqrt(float64(c.n-1)), &pv)
+	return &pv
+}
+
+// Right returns the right-eigenvectors of the canonical correlation matrix if
+// spheredSpace is true, if spheredSpace is false it returns these eigenvectors
+// back-transformed to the original data space.
+func (c CC) Right(spheredSpace bool) (vecs *mat64.Dense) {
+	if c.c == nil || c.y == nil || c.n < 2 {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	var qv mat64.Dense
+	qv.VFromSVD(c.c)
+	if spheredSpace {
+		return &qv
+	}
+
+	var yv mat64.Dense
+	ys := c.y.Values(nil)
+	yv.VFromSVD(c.y)
+
+	scaleColsReciSqrt(&yv, ys)
+
+	qv.Product(&yv, yv.T(), &qv)
+	qv.Scale(math.Sqrt(float64(c.n-1)), &qv)
+	return &qv
+}
+
+// CanonicalCorrelations returns a CC which can provide the results of canonical
+// correlation analysis. The canonical correlation matrix is calculated using
+// the singular value decompositions of the centered input.
+//
+// Canonical correlation analysis can be thought of an approach to finding
+// associations between two sets of variables on the same observations. First,
+// correlations between variables within each set are removed by a 
+// transformation that produces what is called the sphered data in each case,
+// then the linear combinations of the two sets of sphered variables that are 
+// maximally correlated to each other are found.
+//
+// Some notation: let Xc and Yc denote the centered input data matrices x
+// and y (column means subtracted from each column), let Sx and Sy denote the
+// sample covariance matrices within x and y respectively, and let Sxy denote
+// the covariance matrix between x and y. The sphered data can then be expressed
+// as Xc * Sx^(-1/2) and Yc * Sy^(-1/2) respectively, and the correlation matrix
```

Comment:
So basically this gives the solution to the CCA problem I tried to state above.  `Sx^{-1/2}` seems to be the transformation of X (well, Xc), and similarly for Y. If we want to go into the trouble of writing it up, it would be maybe elegant to write first a statement of the CCA problem (a constrained maximization problem like what I wrote above) and this paragraph could describe its solution because it can be found explicitly (as described here).

As a side note, we don't have a consistent notation for matrix transpose, recently I use `^T` in lapack and nobody objects. But I think that we use consistently `{}` instead of `()` in cases like `D^{-1/2}`.